source/agora/consensus/EnrollmentManager.d

```diff
@@ -474,6 +474,16 @@ public class EnrollmentManager
 
     public void clearExpiredValidators (Height block_height) @safe nothrow
     {
+        // clear up the enrollment of a node if the validator cycle of the node
+        // ends at the `block_height`
+        const enrolled = this.validator_set.getEnrolledHeight(this.enroll_key);
+        if (block_height >= enrolled + params.ValidatorCycle)
+        {
+            this.enroll_key = Hash.init;
+            this.data = Enrollment.init;
```

Comment:
What about `signature_noise` and `cycle`? Do they also need to be reset? It might be useful to move this into a new `reset()` function, or named something else more appropriate.